File: dashboard/id_lists.py
"""Parsing of the id lists carried by the observation filter query params.

Every filter that selects a list of ids (species, datasets, areas, basis of
record, data imports) accepts two interchangeable spellings:

    ?speciesIds=1&speciesIds=2&speciesIds=3     one param per id
    ?speciesIds=1-3                             compact

The compact form exists because an alert can select several hundred species,
and one param per id overflows the web server's request-line limit (gunicorn
defaults to 4094 bytes, which a ~400-species alert exceeds). A token is either
a single id or an inclusive `low-high` range, tokens are comma-separated, and
the two spellings may be mixed freely within one request - so the compact form
is purely additive and old clients keep working.

Order and duplicates are preserved: these lists only ever reach the database
through an `IN` clause, so normalizing them here would buy nothing and would
make the parser's output harder to reason about in tests.
"""

from collections.abc import Iterable

# Upper bound on how many ids one filter may expand to. Without it, a
# hand-written `?speciesIds=1-999999999` would have the server build a
# billion-element list on an endpoint that needs no authentication.
MAX_EXPANDED_IDS = 10_000


def parse_id_list(values: Iterable[str]) -> list[int]:
    """Expand the received values of one id filter into a list of ids.

    `values` is the raw list of occurrences of a single query parameter, as
    returned by `QueryDict.getlist()` (each may itself be a comma-separated
    list of tokens).

    Raises ValueError on a malformed token, on a reversed or negative range,
    and when the expansion would exceed MAX_EXPANDED_IDS.
    """
    ids: list[int] = []
    for value in values:
        for token in str(value).split(","):
            token = token.strip()
            if not token:  # `?speciesIds=` and stray commas mean "nothing"
                continue
            ids.extend(_expand_token(token))
            if len(ids) > MAX_EXPANDED_IDS:
                raise ValueError(f"Too many ids requested (limit: {MAX_EXPANDED_IDS})")
    return ids


def _expand_token(token: str) -> Iterable[int]:
    low, separator, high = token.partition("-")
    if not separator:
        return (int(token),)

    # int() rejects the malformed cases for us: an empty bound ("1-", "-5"),
    # a non-numeric one, and a second separator ("1-2-3" -> high == "2-3").
    start = int(low)
    end = int(high)
    if start < 0:
        raise ValueError(f"Negative id in range: {token}")
    if end < start:
        raise ValueError(f"Reversed range: {token}")
    # Measured before expanding, so an absurd range never gets materialized.
    if end - start + 1 > MAX_EXPANDED_IDS:
        raise ValueError(f"Range too wide (limit: {MAX_EXPANDED_IDS}): {token}")
    return range(start, end + 1)

File: assets/frontend/utils/idLists.ts
/**
 * Compact encoding for the id lists carried by the filter params.
 *
 * An alert can select several hundred species, and one query param per id
 * (`speciesIds=1&speciesIds=2&...`) overflows the web server's request-line
 * limit - gunicorn defaults to 4094 bytes, which a ~400-species alert exceeds,
 * and every observation/tile request carries the full filter set.
 *
 * So a list is sent as a single value: comma-separated tokens, where a run of
 * three or more consecutive ids collapses into `low-high`. The backend
 * (dashboard/id_lists.py) accepts this and the historical one-param-per-id
 * spelling interchangeably, so old links keep working.
 */

// Mirrors MAX_EXPANDED_IDS in dashboard/id_lists.py: a hand-written
// `?speciesIds=1-999999999` must not freeze the tab before the request is
// even sent.
const MAX_EXPANDED_IDS = 10000;

// A run of two costs the same either way ("10-11" and "10,11" are both five
// characters), so only three or more is worth a range.
const MIN_RUN_FOR_RANGE = 3;

/**
 * Encode ids as the compact form. The result is sorted and deduplicated, which
 * is what makes runs collapsible and what keeps the address bar canonical: the
 * same selection always produces the same URL. Filter id order is not
 * meaningful anywhere in the app.
 *
 * Returns "" for an empty list, so callers can skip the param entirely.
 */
export function encodeIdList(ids: number[]): string {
    const sorted = [...new Set(ids)].sort((a, b) => a - b);
    const parts: string[] = [];

    let start = 0;
    while (start < sorted.length) {
        let end = start;
        while (end + 1 < sorted.length && sorted[end + 1] === sorted[end] + 1) end++;
        if (end - start + 1 >= MIN_RUN_FOR_RANGE) {
            parts.push(`${sorted[start]}-${sorted[end]}`);
        } else {
            for (let i = start; i <= end; i++) parts.push(String(sorted[i]));
        }
        start = end + 1;
    }

    return parts.join(",");
}

/**
 * Decode either spelling back into ids. Lenient by design - it parses URLs
 * typed or bookmarked by people, so an unparseable token (such as the
 * `areaIds=none` sentinel) is skipped rather than throwing.
 */
export function decodeIdList(raw: string): number[] {
    const ids: number[] = [];

    for (const rawToken of raw.split(",")) {
        const token = rawToken.trim();
        if (!token) continue;

        // indexOf > 0, not >= 0: a leading "-" is a negative number, not a range.
        const separator = token.indexOf("-");
        if (separator > 0) {
            const start = Number(token.slice(0, separator));
            const end = Number(token.slice(separator + 1));
            if (!Number.isInteger(start) || !Number.isInteger(end)) continue;
            if (end < start || end - start + 1 > MAX_EXPANDED_IDS) continue;
            for (let i = start; i <= end; i++) ids.push(i);
        } else {
            const id = Number(token);
            if (Number.isInteger(id)) ids.push(id);
        }
        if (ids.length > MAX_EXPANDED_IDS) return ids.slice(0, MAX_EXPANDED_IDS);
    }

    return ids;
}
